lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1874,51 +2132,177 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
+          // captures in non-escaping closures in Swift 5.7, so in that case we
+          // can only warn until Swift 6.
+          if (AnyFunctionRef(const_cast<AbstractClosureExpr *>(ACE))
+                  .isKnownNoEscape()) {
+            return true;
+          }
+
+          // If this is a capture like `[weak self = somethingElse]`,
+          // this unintentionally permitted implicit self in Swift 5.8 - 5.10,
+          // so we have to only warn until Swift 6.
+          if (auto closureExpr = dyn_cast<ClosureExpr>(ACE)) {
+            if (auto capturedSelfDecl = closureExpr->getCapturedSelfDecl()) {
+              if (closureHasWeakSelfCapture(closureExpr) &&
+                  !isSimpleSelfCapture(capturedSelfDecl)) {
+                return true;
+              }
+            }
+          }
+
+          if (auto conditionalStmt = parentConditionalStmt(selfDecl)) {
+            auto isValidSelfRebinding =
+                hasValidSelfRebinding(conditionalStmt, ACE);
+
+            if (isValidSelfRebinding) {
+              // In Swift 5.8 - 5.10 we unintentionally permitted implicit
+              // self without validating any parent closures. To preserve
+              // source compatibility, in this case we only warn until Swift 6.
+              if (implicitSelfDisallowedDueToInvalidParent(selfDecl, ty, ACE)) {
+                return true;
+              }
+            }
+
+            // In Swift 5.8 - 5.10 we used `requiresLoadExpr` instead
+            // of `requiresCaptureListRef` to validate self unwrap
+            // conditions. For source compatibility, if the legacy check
+            // succeeds but the current check fails, we should only warn
+            // until Swift 6.
+            auto legacyRebindsSelfResult =
+                conditionalStmt->rebindsSelf(ACE->getASTContext(),
+                                             /*requiresCaptureListRef*/ false,
+                                             /*requiresLoadExpr*/ true);
+
+            if (!isValidSelfRebinding && legacyRebindsSelfResult) {
+              return true;
+            }
+          }
+        }
+
+        // If this is a closure nested in a closure with a `[weak self]`
+        // capture, then there is no source-compatibility requirement
+        // to avoid an error.
+        if (auto parentContext = ACE->getParent()) {
+          if (auto parentClosure = dyn_cast_or_null<ClosureExpr>(
+                  parentContext->getInnermostClosureForSelfCapture())) {
+            if (closureHasWeakSelfCapture(parentClosure)) {
+              return false;
+            }
+          }
+        }
+
+        // If the self decl refers to a weak self unwrapping
+        // condition in some parent closure, then there is no
+        // source-compatibility requirement to avoid an error.
+        if (hasValidSelfRebinding(parentConditionalStmt(selfDecl), ACE)) {
+          return false;
         }
```

Comment:
FWIW it looks like the only case that fails with this check removed is:

```swift
    doVoidStuff { [weak self] in      
      guard let self else { return }

      doVoidStuffNonEscaping {
        withEscapingAutoclosure(bar()) // expected-error {{call to method 'bar' in closure requires explicit use of 'self' to make capture semantics explicit}} expected-note {{reference 'self.' explicitly}}
      }
    }
```

But that isn't really to do with the binding, that's because we would have previously referenced the `self` of the function itself. Maybe the carve out should be specific to escaping autoclosures?

Reply:
Good catch! Both of these should be warnings rather than errors. Updated!